### ISSUE TITLE
Consider allowing static constant fields and top level constants to be considered stable and promotable

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#3326]
(Based on https://github.com/dart-lang/sdk/issues/53437#issuecomment-1708434129)

Currently, field promotion (#2020) only considers an expression of the form `target._fieldName` to be stable (and thus promotable) if `_fieldName` represents a promotable field and `target` is one of the following:
- `this` (includes implicit uses of `this`, so bare `_fieldName` is considered promotable inside an instance method of a class)
- `super`
- A local variable that isn't modified or write captured during the time period covered by the potential promotion.
- An expression of the form `target._fieldName` that's considered stable.

If, in addition, we considered a reference to a static constant field or a constant top level variable to be stable, then that would permit additional promotions, for example this code snippet (based on the code snippet in https://github.com/dart-lang/sdk/issues/53437):

```dart
enum E {
  e1, e2;

  final int? _x = 42;
}

main() {
  if (E.e1._x != null) {
    E.e1._x.isEven; // Ok, `E.e1._x` is promoted to non-null
  }
}
```

If we do this, we should probably also consider static constant fields and constant top level variables to be promotable, to preserve the sensible invariant that anything that's stable is promotable (in other words, stability is a stronger notion than promotability). So this would also be allowed:

```dart
const int? i = ...;
main() {
if (i != null) {
  i.isEven; // OK, `i` is promoted to non-null
}
```

Note that in contrast to https://github.com/dart-lang/language/issues/3325, there would be no harm in allowing for this stability and promotability to apply to libraries other than the one declaring the constant, because it's already a breaking change to convert a constant declaration into a non-constant one.

**Comment by lrhn** [comment]
We can consider any constant expression to be stable. Any occurrence of another constant expression with the same value, is the same constant, any promotion applied to that value should still apply.

It's probably not as useful to promote constants in general, as other less constant expressions, because they're quite likely to have their actual type as static type.

**Comment by eernstg** [comment]
Sounds good! I think it's helpful to use the notion of stability (because that's exactly what we need for soundness), and then we can gradually expand on the specified set of stable expressions.

One dichotomy that keeps coming up is this pair of somewhat conflicting perspectives:

1. We should make stability an API property, and it is then safe to rely on stability (e.g., for promotion) globally.
2. So far, stability is an implementation detail, and we should only be allowed to rely on it in the same library.

I agree that it is important to have a clear way to decide whether or not stability is an API property of any given declaration (an "official" property that the maintainers of the declaration have given their commitment). Also, it's certainly reasonable to say that we shouldn't rely on a potentially accidental property globally.

However, I'd very much support going in the direction where stability is an API property: Immutability _is_ of interest to clients.

One way to make stability an API property is to associate some syntax with it (like a keyword `stable`, or adding the modifier `final` on a getter declaration&mdash;anything that unambiguously implies stability). However, we aren't aiming at that kind of syntax changes right now.

However, we could also rely on stability which isn't declared, in situations where the stability property is ensured in the sense that any change (... that we can think of ...) that eliminates the stability would also be a breaking change in some other way. For example, a constant expression can be considered stable because it's already a breaking change in other ways if the relevant declaration(s) are changed to be non-constant.

In other words, I think we can include constant expressions (of any kind) as stable expressions, and remain consistent.

Noticed one typo:

> - An expression of the form `target._fieldName` that's considered stable.

That's an infinite loop, so maybe: "of the form `target._fieldName` where `target` is considered stable."

**Comment by lrhn** [comment]
I suggest the keyword `val`, for "value" because values don't change. And it's shorter than `final`, so people might actually consider to use it.
Being `final` is an implementation details, being `val` is a semantic choice. (And then introduce stable getters, `val int foo => stable-expression;`. The sky's the limit.)

**Comment by stereotype441** [comment]
@lrhn
> Constant expressions are values that we know the _value_ of at compile time. We know everything, including whether it is the same value or not, and any type of has and is safely promotable to. Well, as long as they do not contain any `fromEnvironment` constants.

Even a constant expression that contains a `fromEnvironment` constant should be safe to promote, right? Because the environment can't change during the execution of the program.

**Comment by eernstg** [comment]
> .. **an expression of the form `super._fieldName`, where `_fieldName` represents a promotable instance field.**

This one could be strengthened to "where `_fieldName` represents a final instance field". But we probably don't want to do that, because it wouldn't be good for code readability if everybody starts writing code with `super` all over the place in order to get the promotions. ;-)

**Comment by stereotype441** [comment]
> > .. **an expression of the form `super._fieldName`, where `_fieldName` represents a promotable instance field.**
> 
> This one could be strengthened to "where `_fieldName` represents a final instance field". But we probably don't want to do that, because it wouldn't be good for code readability if everybody starts writing code with `super` all over the place in order to get the promotions. ;-)

Haha, yeah. That's the stuff of nightmares 😃

**Comment by leafpetersen** [comment]
> * Inside an instance method of a class, an expression of the form `super._fieldName`, where `_fieldName` represents a promotable instance field.

Just to be clear, the question of whether `_fieldName` represents a promotable instance field needs to account for a lot if used in a super mixin, right?

> We can consider any constant expression to be stable. Any occurrence of another constant expression with the same value, is the same constant, any promotion applied to that value should still apply.

I'm pretty skeptical that this is worth the cost to implement.

> This one could be strengthened to "where `_fieldName` represents a final instance field". But we probably don't want to do that, because it wouldn't be good for code readability if everybody starts writing code with `super` all over the place in order to get the promotions. ;-)

If users are feeling the pain enough to be willing to do this, why do we feel it's better to make them suffer?

**Comment by stereotype441** [comment]
> > * Inside an instance method of a class, an expression of the form `super._fieldName`, where `_fieldName` represents a promotable instance field.
> 
> Just to be clear, the question of whether `_fieldName` represents a promotable instance field needs to account for a lot if used in a super mixin, right?

Are you responding to my bullet point, or to @eernstg's comment that we could strengthen it to "where `_fieldName` represents a final instance field"?

If you're responding to my bullet point, I don't think we need any logic beyond what we already have for deciding whether `_fieldName` represents a promotable instance field. Namely, it's a promotable instance field if it's final and private, and none of the following exist:
- A non-final instance field with the same name in some other class in the library
- A concrete getter with the same name in some other class in the library
- Some other class in the library that contains a getter with the same name in its interface, but doesn't declare or inherit a concrete implementation of that getter (in other words, a class that contains an implicit `noSuchMethod` forwarding getter with the same name)

We have argued that this is sufficient to make field promotion sound, and I don't see any flaws in our reasoning, even accounting for super mixins.

If you're responding to Erik's comment that we could strengthen it to "where `_fieldName` represents a final instance field", then yes, I think you're right, we can't actually strengthen it like this for the case of super mixins, because the static target of the `super._fieldName` call seen at compile time won't match the runtime target (which gets determined by what superclass the mixin gets applied to). Good catch 😃

> > This one could be strengthened to "where `_fieldName` represents a final instance field". But we probably don't want to do that, because it wouldn't be good for code readability if everybody starts writing code with `super` all over the place in order to get the promotions. ;-)
> 
> If users are feeling the pain enough to be willing to do this, why do we feel it's better to make them suffer?

I think it would be helpful to consider an example of the kind of suffering that Erik's suggestion would address, and what the resulting code would look like. Here's the example that came to mind for me. Let's say a user wants to do this:

```dart
class C {
  final int? i;
  C(this.i);
  foo() {
    if (i != null) {
      // Field promotion means I don't have to do `i!`, right?
      print(i + 1);
    }
  }
}
```
But it doesn't work, because `i` is public, which means it might get overridden in a derived class, so it's not sound to allow it to promote.

If I understand Erik's suggestion correctly, he's saying that if we strengthen the circumstances in which `super.fieldName` is considered promotable (and assuming we finesse the issues around super-mixins), then the user could make their example "work" by doing this:

```dart
class Base { // This class exists so that `i` can be accessed using `super`
  final int? i;
  Base(this.i);
}
class C extends Base {
  C(super.i);
  foo() {
    if (super.i != null) {
      print(super.i + 1);
    }
  }
}
```

My two cents: I know users are feeling the pain of not having field promotion. And when this feature rolls out, they're going to feel the pain of not having field promotion apply to the things they want it to apply to, like public fields. I feel a lot of sympathy to those users. I'm frustrated myself, that we had to narrow the scope of field promotion to private final fields in order to make it sound, and I'm really hopeful that we can make future language improvements to broaden the set of fields that are promotable.

But I personally doubt that users are feeling the pain of not having field promotion so badly that they're going to want to do a hack like the one above. And even if I'm wrong, I would _much_ rather fast-track some other language feature that gets them what they need without having to make such an ugly hack, like for example:
- Do some analysis of the class hierarchy within the library, taking into account that some classes are private, and others are marked `sealed` or `final`, and use that to conclude that some public final fields are safe to promote due to the fact that there's no way for a class outside the library to override or implement them. Of course, there would be a lot of details to sort out to make sure the analysis is sound, accounting for the fact that class modifiers like `final` can be ignored within a library. If we do this, the user could get type promotion in the example above by marking class `C` as `final`.
- As in the previous bullet, but also conclude that some public final fields are safe to promote when accessed through `this`, due to the fact that there's no way for a class outside the library to override them. If we do this, the user could get type promotion in the example above by marking class `C` as `interface`.
- Add a way for a user to mark a field as "not overridable", so that the compiler can prove that the reference to `this.i` really does get dispatched to the final field at runtime. Again, there would be details to sort out. Like: is this the same as saying that the field is "not virtual"? Can you _implement_ a class with a non-overridable field? Assuming we sort out those details, our hypothetical user would get the type promotion they want by marking `i` as "not overridable".
- Some variant of the stable getters proposal, in which case `i` would be safe to promote because we know statically that any override of it must necessarily return a stable value.
